roi_extractors: GRoIE addition

hachreak opened the pull request:
Signed-off-by: Leonardo Rossi <leonardo.rossi@studenti.unipr.it>

ZwwWayne commented on configs/groie/mask_rcnn_r50_fpn_1x-orig.py:
```diff
@@ -0,0 +1,190 @@
+# model settings
```
This config is not necessary to be kept since it is the same as the original file.

ZwwWayne commented on mmdet/models/roi_extractors/groie.py:
```diff
@@ -0,0 +1,162 @@
+    def __init__(self, **kwargs):
+        kwargs = deepcopy(kwargs)
+
+        post_conf = kwargs.pop('post_conf')
```
`post_module` or `post_cfg` fits the codebase's convention better, so as `pre_module` or `pre_cfg`.

ZwwWayne commented on mmdet/models/roi_extractors/groie.py:
```diff
@@ -0,0 +1,162 @@
+        super(SumGenericRoiExtractor, self).__init__(**kwargs)
+
+        # build pre/post processing modules
+        self.post_conv = load_processing_model(post_type, post_conf)
```
`slef.post_module` might be more suitable since sometimes it is not a con layer.

hachreak commented:
Hi @ZwwWayne, thanks a lot for reviewing my code!
I changed the code as you mentioned.
I think there is an issue about python 3.5 because `mmcv` doesn't support it anymore.

ZwwWayne commented:
> Hi @ZwwWayne, thanks a lot for reviewing my code!
> I changed the code as you mentioned.
> I think there is an issue about python 3.5 because `mmcv` doesn't support it anymore.

Hi @hachreak ,
Thanks for your kind contribution. We just release MMDetection V2.0 with faster training speed and better performance, which no longer supports python 3.5 anymore and support pytorch>=1.3.

Would you like to migrate your PR to the MMDetection V2.0 version?

hachreak commented:
> Would you like to migrate your PR to the MMDetection V2.0 version?

I'll do! 
I need first to install python 3.7 on my debian and see how refactor the code to be usable on mmdetection v2.
Any help is welcome! 
Thanks a lot :smile:

ZwwWayne commented on .gitignore:
```diff
@@ -119,3 +119,6 @@ work_dirs/
 *.sh~
+
+# dataset
+data/
```
line 107 already adds `./data/`, it should already work?

hachreak replied:
I didn't noticed line 107, at least in my environment git (v2.11.0) complains and tell me `data` directory is part of untracked files.
It looks like the `./`has confused git in my case.
But removing initial `./` works for me! :smile:

ZwwWayne replied:
OK, we may keep both for now.

ZwwWayne commented on configs/grid_rcnn/grid_rcnn_r50_fpn_gn-head_1x_coco.py:
```diff
@@ -0,0 +1,11 @@
+    policy='step',
+    warmup='linear',
+    warmup_iters=500,
+    warmup_ratio=1.0 / 3,
```
0.001 is more commonly used in MMDet V2.0. The difference is marginal but it can stabilize training.

hachreak replied:
done. thanks :smile:

ZwwWayne commented on configs/groie/README.md:
```diff
@@ -0,0 +1,65 @@
+
+Therefore, the proposed layer (called Generic RoI Extractor - **GRoIE**)
+introduces non-local building blocks and attention mechanisms to boost the
+performance.
```
You can add the Bibtex here like other config directories for the convenience of citation.

ZwwWayne commented on configs/groie/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## Contact
+
+This repo is currently maintained by
```
Would `The implementation of GROI is currently...` be more appropriate?

hachreak replied:
Right, thanks :smile:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,163 @@
+@ROI_EXTRACTORS.register_module
+class SumGenericRoiExtractor(SingleRoIExtractor):
+    """Extract RoI features from all summed feature maps levels.
+
```
Adding the paper link here might be better since it will appear in the API documentation.

hachreak replied:
done, thanks :smile:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,163 @@
+        ... (see SingleRoIExtractor for other parameters)
+    """
+
+    def __init__(self, **kwargs):
```
Following the MMDetection's code-style, it might be more consistent to put new arguments here, like 
```python
def __init__(self, post_cfg,  pre_cfg, reduce_cfg, **kwargs):
```
This can also make the API documentation more clear.

hachreak replied:
done. thanks :smile:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,163 @@
+        ... (see SingleRoIExtractor for other parameters)
+    """
+
+    def __init__(self, **kwargs):
```
See commonts in line 118.

hachreak replied:
done. Thanks :smile:

ZwwWayne commented on configs/groie/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## Results and models
+
+The results on COCO 2017v minival (5k images) are shown in the below table.
```
We usually simply use COCO 2017 rather than COCO 2017v

hachreak replied:
I think it's a typo here: I mean `COCO 2017`

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,151 @@
+
+def load_processing_model(model_type, config):
+    """Load a specific module."""
+    return models.get(model_type, NoProcess)(**config)
```
`NoProcess` might not be necessary, `nn.Sequential()` could do the same thing?

hachreak replied:
I added it inside the plugin list instead of use `NoProcess` class :+1:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,151 @@
+    'GeneralizedAttention': GeneralizedAttention,
+    'ConvModule': ConvModule,
+    'Sequential': SequentialModule,
+    'ReLU': nn.ReLU,
```
1. From the configs, it seems that 'ReLU' and `Sequential` are hardly used, do we need to keep them?
2. MMDetection V2.0 already supports [plugin](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/ops/plugin.py#L9), and it seems simply adding `ConvModule` into plugin could meet the needs in GROIE.

hachreak replied:
1. Could you explain what you mean? thanks :smile: 
2. I'm rewriting this part to use plugins! :+1:

ZwwWayne replied:
For the first point, I mean that it seems `Sequential` or `NoProcess` seems to be not used in post_module or pre_modules. Therefore, it seems that we do not need to keep them as an option.

hachreak replied:
removed. :smile:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,110 @@
+        super(SumGenericRoiExtractor, self).__init__(**kwargs)
+
+        # build pre/post processing modules
+        _, self.post_module = build_plugin_layer(post_cfg, '_post_module')
```
```python
self.post_module = build_plugin_layer(post_cfg, '_post_module')[1]
```

hachreak replied:
done :smile:

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,110 @@
+        super(ConcatGenericRoiExtractor, self).__init__(**kwargs)
+
+        # build pre/reduce/post processing modules
+        _, self.post_module = build_plugin_layer(post_cfg)
```
```python
self.post_module = build_plugin_layer(post_cfg)[1]
```

hachreak replied:
done

ZwwWayne commented on mmdet/models/roi_heads/roi_extractors/groie.py:
```diff
@@ -0,0 +1,110 @@
+
+        for i in range(num_levels):
+            # apply pre-processing to a RoI extracted from each layer
+            rois_ = rois
```
Maybe we can directly use `rois` rather than `rois_`?

hachreak replied:
done

ZwwWayne commented:
1. The Master branch already fixes the bug in CI. Please merge the master branch thus the CI could perform normally.
2. Since MMDetection V2.0, to reduce the probability of bugs caused by code development, we suggest developers also prepare unittests for new modules, e.g., two unit tests for `SumGenericRoiExtractor` and `ConcatGenericRoiExtractor` to check whether they could be initialized as what we expect and whether they could handle some illegal input.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmdetection?pullRequest=2584) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmdetection?pullRequest=2584) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmdetection?pullRequest=2584) it.</sub>

ZwwWayne commented:
We can also add a link in [model zoo](https://github.com/open-mmlab/mmdetection/blob/master/docs/model_zoo.md).

hachreak commented:
> We can also add a link in [model zoo](https://github.com/open-mmlab/mmdetection/blob/master/docs/model_zoo.md).

Added and rebased to master! :smile:

hellock commented:
@ZwwWayne We need to benchmark Grid R-CNN R50 1x, and also host the GRoIE checkpoints.

ZwwWayne commented:
> @ZwwWayne We need to benchmark Grid R-CNN R50 1x, and also host the GRoIE checkpoints.

Sure.
Hi @hachreak ,
Do you have any way to host your checkpoints and logs? For example, the google drive or any other way that make the checkpoints and logs publicly accessible.
Or you can find a way to share me your trained checkpoints and their logs, I can help to upload them to the open-mmlab's AWS server.

hachreak commented:
> Do you have any way to host your checkpoints and logs?

The only checkpoints I have are for a quite old version of mmdetection.

hellock commented:
> > Do you have any way to host your checkpoints and logs?
> 
> The only checkpoints I have are for a quite old version of mmdetection.

Then we will help benchmark the models.

hachreak commented:
> Then we will help benchmark the models.

Hi @hellock can I know what will happen to this PR? thanks

ZwwWayne commented:
> > Then we will help benchmark the models.
> 
> Hi @hellock can I know what will happen to this PR? thanks

Do not worry, we are benchmarking the models now.

hellock commented:
> > Then we will help benchmark the models.
> 
> Hi @hellock can I know what will happen to this PR? thanks

Sorry I accidentally closed this PR, it is reopened.

ZwwWayne commented:
Hi @hachreak ,
Your model is re-benchmarked. Because I could not push commits to branch in your repo, I pushed them to mine. Could your modify your branch according to the modification in this [branch](https://github.com/ZwwWayne/mmdetection/commit/3543df74221374bb8f30a46e017446bb20ecfeef)? The performance is updated and a bug is fixed in empirical attention caused by empty tensor.

hachreak commented:
Hi @ZwwWayne 
thanks a lot! I updated the branch with your commit and rebased on top of master! :smile:

OpenMMLab-Coodinator commented:
Hi @hachreak! First of all, we want to express our gratitude for your significant PR in the MMXXX project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
